rust/datafusion/src/datasource/datasource.rs

```diff
@@ -0,0 +1,46 @@
+// Licensed to the Apache Software Foundation (ASF) under one
+// or more contributor license agreements.  See the NOTICE file
+// distributed with this work for additional information
+// regarding copyright ownership.  The ASF licenses this file
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License.  You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+//! Data source traits
+
+use std::cell::RefCell;
+use std::rc::Rc;
+use std::sync::Arc;
+
+use arrow::datatypes::Schema;
+use arrow::record_batch::RecordBatch;
+
+use crate::execution::error::Result;
+
+/// Source table
+pub trait Table {
+    /// Get a reference to the schema for this table
+    fn schema(&self) -> &Arc<Schema>;
+
+    /// Perform a scan of a table and return an iterator over the data
+    fn scan(
+        &self,
+        projection: &Option<Vec<usize>>,
+        batch_size: usize,
+    ) -> Rc<RefCell<RecordBatchIterator>>;
+}
+
+/// Iterator for reading a series of record batches with a known schema
+pub trait RecordBatchIterator {
```

Comment:
@liurenjie1024 might this be a good starting place for [ARROW-3706](https://issues.apache.org/jira/browse/ARROW-3706)?

Reply:
Yes.  I would suggest it extends Iterator<Item=Result<RecordBatch>> to be rust idomic.

Reply:
I have started experimenting with support for parallel execution with threads and we are going to have to revisit some of the design to support this so i think it would be good to work together on this. I was thinking about creating a PR to show some ideas (separate from the rest of the project). I hope to get to that in the next week or two.